secret-rotator/rotator/rotator.go

```diff
@@ -0,0 +1,263 @@
+// Returns error if fails.
+func (r *SecretRotator) UpsertLabels(rotatedSecret config.RotatedSecretSpec) error {
+	_, err := r.Client.GetSecretLabels(rotatedSecret.Project, rotatedSecret.Secret)
+	if err != nil {
+		return err
+	}
+
+	// attach the labels needed for the provisioner
+	for key, val := range rotatedSecret.Type.Labels() {
+		err = r.Client.UpsertSecretLabel(rotatedSecret.Project, rotatedSecret.Secret, key, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Refresh checks if the secret needs to be refreshed, and if so
+// provisions a new secret and updates the Secret Manager secret.
+// Returns true if the secret is refreshed.
+func (r *SecretRotator) Refresh(rotatedSecret config.RotatedSecretSpec, now time.Time) (bool, error) {
+	shouldRefresh, err := r.ShouldRefresh(rotatedSecret, now)
+	if err != nil {
+		return false, err
+	}
+
+	if !shouldRefresh {
+		return false, nil
+	}
+
+	labels, err := r.Client.GetSecretLabels(rotatedSecret.Project, rotatedSecret.Secret)
+	if err != nil {
+		return false, err
+	}
+
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	// attach the labels needed for the provisioner
+	for key, val := range rotatedSecret.Type.Labels() {
+		labels[key] = val
+	}
+
+	newId, newSecret, err := r.Provisioners[rotatedSecret.Type.Type()].CreateNew(labels)
+	if err != nil {
+		return false, err
+	}
+
+	// update the secret Manager secret
+	latestVersion, err := r.Client.UpsertSecret(rotatedSecret.Project, rotatedSecret.Secret, newSecret)
+	if err != nil {
+		return false, err
+	}
+
+	// keys in format of "v%d" indicate that they are (version: id) pairs attached by the rotator
+	// the reason for the prefix "v" is that Secret Manager labels need to begin with a lowwer case letter
+	err = r.Client.UpsertSecretLabel(rotatedSecret.Project, rotatedSecret.Secret, "v"+latestVersion, newId)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+
+}
+
+// ShouldRefresh checks whether the secret needs to be refreshed according to 'now' and 'rotatedSecret.Refresh'.
+// Returns true if the secret needs to be refreshed.
+func (r *SecretRotator) ShouldRefresh(rotatedSecret config.RotatedSecretSpec, now time.Time) (bool, error) {
+	err := r.Client.ValidateSecretVersion(rotatedSecret.Project, rotatedSecret.Secret, "1")
+	if err != nil {
+		// create the secret and/or dirst version if it does not already exist
+		if status.Code(err) == codes.NotFound {
+			return true, nil
+		}
+		return false, err
+	}
+
+	createTime, err := r.Client.GetCreateTime(rotatedSecret.Project, rotatedSecret.Secret, "latest")
+	if err != nil {
+		return false, err
+	}
+
+	if rotatedSecret.Refresh.Interval != 0 {
+		// the refresh stratetgy is refreshInterval
+		// check the elapsed time from its createTime to now.
+		if now.After(createTime.Add(rotatedSecret.Refresh.Interval)) {
+			return true, nil
+		}
+	} else {
+		// the refresh strategy is cron
+		// TODO
```

Comment:
I thibk I will drop cron entirely for now, but keeping the current `RotatedSecretConfig` structure so that there's still room to support cron in the future.